cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean

```diff
@@ -0,0 +1,1628 @@
+import Cedar.Thm.Validation.WellTyped.Definition
+import Cedar.Thm.SymCC.Compiler.WF
+import Cedar.Thm.SymCC.Env.ofEnv
+import Cedar.Thm.SymCC.Env.WF
+
+/-!
+This file contains theorems saying that `compile` succeeds
+with the correct term type on well-typed expressions.
+-/
+
+namespace Cedar.Thm
+
+open Cedar.Spec
+open Cedar.Thm
+open Cedar.Validation
+open SymCC
+
+/--
+States the symbolic compiler succeeds on a typed expression `tx` and
+produces a term `t` with the corresponding type.
+-/
+def CompileWellTyped (tx : TypedExpr) (εnv : SymEnv) : Prop :=
+  ∃ t : Term,
+    compile tx.toExpr εnv = .ok t ∧
+    t.typeOf = .option (TermType.ofType tx.typeOf)
+
+/--
+A sufficient condition for `CompileWellTyped` to hold.
+-/
+def CompileWellTypedCondition (tx : TypedExpr) (Γ : Environment) (εnv : SymEnv) : Prop :=
+  εnv = SymEnv.ofEnv Γ ∧
+  TypedExpr.WellTyped Γ tx ∧
+  εnv.WellFormedFor tx.toExpr
```

Comment:
Could this be refactored to not require `εnv` as a parameter, and replace all usages of it with `SymEnv.ofEnv Γ`?  This would improve ergonomics and I think clarity?  (Open to arguments why this would not improve clarity)

Reply:
I was trying this change, but I realized that a lot of the theorems have the `εnv` parameter and rephrasing them in term of `SymEnv.ofEnv Γ` might make them a bit complicated.

Also this definition is likely subject to change once we prove `WellFormedFor` from the first two conditions. So maybe we can simplify things then?

Reply:
Sounds good to me.